### PR TITLE
gcc: add patch for sanitizer in gcc<11.1.0

### PR DESCRIPTION
**Pull request by acastanedam**
This allows to fix the compilation of gcc versions less than 11.1.0,
due to the remove of cyclades of libsanitizer as it is described in
the patch file.

### CODE REVIEW

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/gcc/package.py]: Would it be possible to fetch this patch from some url? Also, are all these patches modifying the same behavior across different GCC versions?

Comment (acastanedam) [reply var/spack/repos/builtin/packages/gcc/package.py]: Hi @alalazo, in this address there is a link for the patch (https://gcc.gnu.org/git/gitweb.cgi?p=gcc.git;h=2b40941d23b1570cdd90083b58fa0f66aa58c86e), but I do not know, if it is permanent. In regard to the second question, I can not assert for all the versions available in Spack, I have checked it with three different versions < 11.1.0. Do we need some more test cases?

Comment (alalazo) [reply var/spack/repos/builtin/packages/gcc/package.py]: Let's use this url or something similar:
- https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff;h=2b40941d23b1570cdd90083b58fa0f66aa58c86e;hp=e9aeff6adbd8655ef3165cada15fda5fb6267551

and delete the patch from Spack repo. Even if it is not permanent, we'll cache it in our source mirror when it gets updated. 

>  In regard to the second question, I can not assert for all the versions available in Spack, I have checked it with three different versions < 11.1.0. Do we need some more test cases?

The question is just to understand if all the patch directives called `glibc-2.31-libsanitizer-*` do the same thing to different versions of GCC (I admittedly didn't take the time to have a look at that myself, sorry). If not, please put a brief comment above this patch to explain what it changes.

Comment (acastanedam) [reply var/spack/repos/builtin/packages/gcc/package.py]: Hi,
I already did it. One question, should I create a new  merge request or there is a way to push the new commit to this MR?

Comment (alalazo) [reply var/spack/repos/builtin/packages/gcc/package.py]: @acastanedam The PR is fine like it is, thanks. I'll be waiting for tests to finish and merge it.

**Review by alalazo — APPROVED** [review]

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (acastanedam) [reply var/spack/repos/builtin/packages/gcc/package.py]: Alright, thanks to you

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/gcc/package.py]: Did you try the patch with older versions of GCC? I wanted to double check a couple before merging and this is what I get:
```console
$ spack patch gcc@:10
==> Using cached archive: /home/culpo/PycharmProjects/spack/var/spack/cache/_source-cache/archive/64/64f404c1a650f27fc33da242e1f2df54952e3963a49e06e73f6940f3223ac344.tar.xz
==> Using cached archive: /home/culpo/PycharmProjects/spack/var/spack/cache/_source-cache/archive/b4/b48e48736062e64a6da7cbe7e21a6c1c89422d1f49ef547c73b479a3f3f4935f
The text leading up to this was:
--------------------------
|diff --git a/libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.cc b/libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.cc
|index d823a12190c..e8fce8a0287 100644
|--- a/libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.cc
|+++ b/libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.cc
--------------------------
File to patch: X
X: No such file or directory
Skip this patch? [y] y
3 out of 3 hunks ignored
2 out of 2 hunks FAILED -- saving rejects to file libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.h.rej
==> Patch https://gcc.gnu.org/git/?p=gcc.git;a=patch;h=2b40941d23b1570cdd90083b58fa0f66aa58c86e failed.
==> Error: Command exited with status 1:
'/usr/bin/patch' '-s' '-p' '1' '-i' '/tmp/culpo/spack-stage/spack-stage-ou1_9fpk/?p=gcc.git;a=patch;h=2b40941d23b1570cdd90083b58fa0f66aa58c86e' '-d' '.'
```

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (acastanedam) [reply var/spack/repos/builtin/packages/gcc/package.py]: Yes, gcc 7.5.0 and gcc 8.5.0. If you want, we can try more versions

Comment (alalazo) [reply var/spack/repos/builtin/packages/gcc/package.py]: I think we need to split this patch in two. The issue is that `sanitizer_platform_limits_posix.cc` got renamed at some point to `sanitizer_platform_limits_posix.cpp`. For later GCC you can use this:
```python
patch('https://github.com/gcc-mirror/gcc/commit/745dae5923aba02982563481d75a21595df22ff8.patch', sha256='7686d3b3a7ad0dc3f44a4141067133d2c708ea2cba503f684389e7bbb3bfa81a', when='@:11.1.0')
```
modulo the lower bound for versions.

Comment (acastanedam) [reply var/spack/repos/builtin/packages/gcc/package.py]: Hi @alalazo 

I have just pushed the changes taking into account the actual extension for the versions less (greater) or equal than 9.30 (10.1.0), such that `spack patch` works. The only thing is strange is the case with version 9.4.0 which does not work with either of the patches.

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py]: This is missing a lower version. The earliest it seems to apply for is 6.5.0. Please test `spack patch` for *all* versions between 6.5.0 and 9.3.0.

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py]: Any reason you are not using https://gcc.gnu.org/git/?p=gcc.git;a=patch;h=745dae5923aba02982563481d75a21595df22ff8 here (to be consistent with the first patch)?

**Review by michaelkuhn — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/gcc/package.py]: Might be my fault as I was suggesting the second patch above. But I agree to be consistent with the two patches.

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (acastanedam) [reply var/spack/repos/builtin/packages/gcc/package.py]: Hi @alalazo, @michaelkuhn

I have updated the script with the requested changes. I checked with `spack patch`, as well as, compiling some specific versions requiring (or not) either of the patches.

File: var/spack/repos/builtin/packages/gcc/package.py
Comment (acastanedam) [reply var/spack/repos/builtin/packages/gcc/package.py]: Hi @alalazo, @michaelkuhn

I have updated the script with the requested changes. I checked with spack patch, as well as, compiling some specific versions requiring (or not) either of the patches.

Comment (acastanedam) [reply var/spack/repos/builtin/packages/gcc/package.py]: Hi,

Sorry, I have to apologize. I have just realized versions like 7.1.0 or 8.1.0 do actually need the patch for compilation. The problem is the patching process fails, e.g., for gcc@7.1.0:
`==> Fetching https://mirror.spack.io/_source-cache/archive/8a/8a8136c235f64c6fef69cac0d73a46a1a09bb250776a050aec8f9fc880bebc17.tar.bz2
==> Fetching https://gcc.gnu.org/git/?p=gcc.git;a=patch;h=2b40941d23b1570cdd90083b58fa0f66aa58c86e
==> Applied patch /opt/spack/var/spack/repos/builtin/packages/gcc/glibc-2.31-libsanitizer-1.patch
==> Applied patch /opt/spack/var/spack/repos/builtin/packages/gcc/glibc-2.31-libsanitizer-2-gcc-7.patch
1 out of 3 hunks FAILED -- saving rejects to file libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.cc.rej
==> Patch https://gcc.gnu.org/git/?p=gcc.git;a=patch;h=2b40941d23b1570cdd90083b58fa0f66aa58c86e failed.
==> Error: Command exited with status 1:
'/usr/bin/patch' '-s' '-p' '1' '-i' '/home/castanedamedina/.spack/stage/spack-stage-_q0btyzf/?p=gcc.git;a=patch;h=2b40941d23b1570cdd90083b58fa0f66aa58c86e' '-d' '.'
`
Probably one needs a third patch file for such cases.

Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/gcc/package.py]: I have seen similar problems with GCC patches in the past since even minor releases seem to move stuff around. I would say the two patches you added are good enough, people will hopefully not use very old versions of GCC on very new versions of Linux.

**Review by michaelkuhn — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
Thanks @acastanedam !

**Pull request merged**
